File: Tang/Tang/Content/Module/VideoPlayer/ViewCtrl/TangPlayerViewCtrl.m
//
//  TangPlayerViewCtrl.m
//  Tang
//
//  Created by wwwbbat on 2017/9/17.
//  Copyright © 2017年 tiny. All rights reserved.
//

#import "TangPlayerViewCtrl.h"
#import "TangPlayerView.h"
#import "TangPostVideoParser.h"
@import AVFoundation;
@import ReactiveObjC;
#import <KTVHTTPCache/KTVHTTPCache.h>

@interface TangPlayerViewCtrl ()
@property (weak, nonatomic) IBOutlet TangPlayerView *playerView;
@property (weak, nonatomic) IBOutlet UIView *coverView;


@property (weak, nonatomic) IBOutlet UIView *controlView;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *loadingActivity;
@property (weak, nonatomic) IBOutlet UISlider *sliderBar;
@property (weak, nonatomic) IBOutlet UIProgressView *cacheProgressBar;
@property (weak, nonatomic) IBOutlet UILabel *countdownLabel;

@property (strong, nonatomic) AVPlayer *player;

@end

@implementation TangPlayerViewCtrl

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    [self initUI];
    
    [self play];
}

- (void)initUI
{
    UIImage *image = [UIImage imageWithColor:[UIColor whiteColor] size:CGSizeMake(16, 16)];
    image = [image imageByRoundCornerRadius:image.size.width/2];
    [self.sliderBar setThumbImage:image forState:UIControlStateNormal];
    [self.sliderBar setThumbImage:image forState:UIControlStateHighlighted];
    [self.sliderBar setThumbImage:image forState:UIControlStateSelected];
}

- (void)play
{
    [SVProgressHUD show];
    RunOnGlobalQueue(^{
        TangVideoBaseData *data = [TangPostVideoParser parsePost:self.post];
        RunOnMainQueue(^{
            [SVProgressHUD dismiss];
            if (data) {
                
                NSString *proxyURL = [KTVHTTPCache proxyURLStringWithOriginalURLString:data.video];
                self.player = [AVPlayer playerWithURL:[NSURL URLWithString:proxyURL]];
                self.playerView.playerLayer.player = self.player;
                [self addObserver];
                [self.player play];

            }else{
                [SVProgressHUD showErrorWithStatus:@"播放失败"];
            }
        });
    });
}

- (void)addObserver
{
    
//    ygweakify(self);
//    [RACObserve(self.player, duration)
//     subscribeNext:^(id x) {
//         ygstrongify(self);
//         NSInteger remain = self.player.duration * self.player.progress;
//         self.countdownLabel.text = [NSString stringWithFormat:@"%02d:%02d",remain/60,remain%60];
//     }];
//    [RACObserve(self.player, progress)
//     subscribeNext:^(id x) {
//         ygstrongify(self);
//         self.sliderBar.value = self.player.progress;
//         NSInteger remain = self.player.duration * self.player.progress;
//         self.countdownLabel.text = [NSString stringWithFormat:@"%02d:%02d",remain/60,remain%60];
//     }];
}

- (IBAction)exit:(id)sender
{
//    [self.player stop];
    self.player = nil;
    [self doLeftNaviBarItemAction];
}

- (IBAction)playTimeControl:(id)sender
{
    
}



@end

File: Tang/Tang/AppCode/Module/Player/Ctrls/TangPlayerViewCtrl.swift
//
//  TangPlayerViewCtrl.swift
//  Tang
//
//  Created by Tiny on 2018/2/9.
//  Copyright © 2018年 tiny. All rights reserved.
//

import UIKit
import AVFoundation
import SVProgressHUD
import KTVHTTPCache
import IJKMediaFramework
import ReactiveCocoa
import ReactiveSwift

class TangPlayerViewCtrl: YGBaseViewCtrl {
    
    @IBOutlet weak var playerContainer: TangPlayerView!
    @IBOutlet weak var previewView: TangPreviewView!
    @IBOutlet weak var coverView : UIView!
    @IBOutlet weak var controlView : TangPlayerControl!
    
    private var player : TangPlayer!
    private var playerController : IJKAVMoviePlayerController!
    private var playerView : UIView!
    
    var data : TangPlayerSharedData!
    private var post : TangVideoPost {
        get {
            return data.post
        }
    }
    private var info : TangVideoInfo? {
        get {
            return data.info
        }
    }
    private var autoPlayFlag = false
    
    override func viewDidLoad() {
        super.viewDidLoad()

        initUI()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        if !autoPlayFlag {
            autoPlayFlag = true
            prepare()
        }
    }
    
    deinit {
        
    }

    private func initUI() {
        
        self.previewView.show(data.focusImage)
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
    }
    
    private func prepare() {
        SVProgressHUD.show()
        data.getVideoInfo { [unowned self] (info) in
            SVProgressHUD.dismiss()
            self.initPlayer()
        }
    }
    
    private func initPlayer() {
        if let info = self.info, let video = info.video {
            let proxyURL = KTVHTTPCache.proxyURLString(withOriginalURLString: video)
            print("Play video : \(video), proxyURL : \(proxyURL!) ")
            self.player = TangPlayer.init(url: proxyURL!, layer: self.playerContainer.playerLayer, delegate: self)
        }else{
            SVProgressHUD.showError(withStatus: "播放失败")
        }
    }
    
    @IBAction func exit(_ sender : Any) {
        self.player = nil
        doLeftNaviBarItemAction()
    }
}

extension TangPlayerViewCtrl : TangPlayerCallback {

    func playerLayerIsReady(_ player: TangPlayer) {
        self.previewView.isHidden = true
    }
    
    func player(_ player: TangPlayer, time: Float, total: Float) {
        controlView.update(time, total)
    }
    
    func player(_ player: TangPlayer, state: TangPlayerState) {
        print("Player state : ",state)
    }
    
    func playerWillPlaying(_ player: TangPlayer, rate: Float) {
        
    }
    
    func playerDidPaused(_ player: TangPlayer) {
        
    }
}

extension TangPlayerViewCtrl : TangPlayerControlCallback {
    
    func playerControl(_ control: TangPlayerControl, slider state: TangPlayerControlSlidingState, value: Float) {
        switch state {
        case .begin:
            player.pause()
            updateWillSeekHUD(seek: player.curTime)
        case .update:
            updateWillSeekHUD(seek: player.duration * Float64(value))
            break
        case .end:
            SVProgressHUD.dismiss()
            player.seek(progress: value)
        case .cancel:
            SVProgressHUD.dismiss()
        }
    }
    
    func playerControl(_ control: TangPlayerControl, changePlayback state: TangPlaybackState) {
        switch state {
        case .idle: break
        case .loading:  break
        case .playing:
            player.play()
        case .pause:
            player.pause()
        case .end:  break
        }
    }
    
    func updateWillSeekHUD(seek to : Float64) {
        let text = String.readableTimeInterval(to) + " / " + String.readableTimeInterval(player.duration)
        SVProgressHUD.show(text)
    }
}
